respy/simulate.py

```diff
@@ -226,84 +217,80 @@ def _extend_data_with_sampled_characteristics(df, optim_paras, options):
         A pandas DataFrame with no missings at all.
 
     """
-    index = df.index
+    # Sample characteristics only for the first period.
+    fp = df.query("period == 0").copy()
+    index = fp.index
 
     for observable in optim_paras["observables"]:
         level_dict = optim_paras["observables"][observable]
-        sampled_char = _sample_characteristic(df, options, level_dict, use_keys=False)
-        df[observable] = df[observable].fillna(
+        sampled_char = _sample_characteristic(fp, options, level_dict, use_keys=False)
+        fp[observable] = fp[observable].fillna(
             pd.Series(data=sampled_char, index=index), downcast="infer"
         )
 
     for choice in optim_paras["choices_w_exp"]:
         level_dict = optim_paras["choices"][choice]["start"]
-        sampled_char = _sample_characteristic(df, options, level_dict, use_keys=True)
-        df[f"exp_{choice}"] = df[f"exp_{choice}"].fillna(
+        sampled_char = _sample_characteristic(fp, options, level_dict, use_keys=True)
+        fp[f"exp_{choice}"] = fp[f"exp_{choice}"].fillna(
             pd.Series(data=sampled_char, index=index), downcast="infer"
         )
 
     for lag in reversed(range(1, optim_paras["n_lagged_choices"] + 1)):
         level_dict = optim_paras[f"lagged_choice_{lag}"]
-        sampled_char = _sample_characteristic(df, options, level_dict, use_keys=False)
-        df[f"lagged_choice_{lag}"] = df[f"lagged_choice_{lag}"].fillna(
+        sampled_char = _sample_characteristic(fp, options, level_dict, use_keys=False)
```

Comment:
Dont we need use_keys=True here. I am getting an error because sample characteristics does not return categories.

Reply:
This is intended. Since https://github.com/OpenSourceEconomics/respy/pull/284 respy does not use categoricals for choices and obervables. Internally, everything is converted to codes.